File: Competitive Codes/ESKO_2020/esko.cpp
#include <bits/stdc++.h>
#include <sstream>
using namespace std;
unordered_map<string,int> mp;
unordered_map<int,string> rmp;
unordered_map<int,int> mp_pro;
struct node{
    int dest;
    int weight;
    bool reverse;
};
void addedge(vector<node> adj[],int s,int e,int val){
    node save;
    save.dest = e;
    save.weight = val;
    save.reverse = false;
    adj[s].push_back(save);
    save.dest = s;
    save.reverse = true;
    adj[e].push_back(save);
}
vector<string> splitter(string ss){
    vector<string> input;
	int pos = 0, word = 0,len=0;
	for(int i=0;i<ss.size();i++){
	    len++;
	    if(ss[i]==' '){
	       input.push_back(ss.substr(pos,len-1));
	       len=0;
	       pos=i+1;
	    }
	}
	return input;
}

void dfs(vector<node> adj[],int s,vector<bool>& visited,vector<string>& result){
    visited[s] = true;
    for(auto noder:adj[s]){
        if(!visited[noder.dest]){
            
            if(noder.reverse) mp_pro[noder.dest] =mp_pro[s] / noder.weight;
            else mp_pro[noder.dest] =mp_pro[s] * noder.weight;
           
            ostringstream numtostring;
            numtostring<<mp_pro[noder.dest];
            string save ="";
            save += numtostring.str();
            save += rmp[noder.dest];
            result.push_back(save);
            dfs(adj,noder.dest,visited,result);
        }
    }
}

int main(){
    int n;
    cin>>n;
    string point[n];
    int kk = 0;
    int savo = n;
    while(savo--){
        cin>>point[kk];
        kk++;
    }
    
    int k=0;
    for(auto i:point) mp[i] = k++;
    k=0;
    for(int i=0;i<n;i++) rmp[i] = point[k++];
    vector<node> adj[n];
    int fill_nodes = n;
    int leave = 0;
    while(fill_nodes--){

        string s;
        getline(cin,s);
        if(leave==0){leave++;continue;}
        s+=" ";
        
        vector<string> ss = splitter(s);
        stringstream geek(ss[2]); 
        int val = 0; 
        geek >> val;
        
        addedge(adj,mp[ss[0]],mp[ss[3]],val);
    }
    
    for(int i=0;i<n;i++){
        vector<bool> visited(n,false);
        for(int i=0;i<n;i++) mp_pro[i] = 1;
        vector<string> result;
        dfs(adj,i,visited,result);
        sort(result.begin(),result.end());
        if(result[0][0]!='0'){
            cout<<"1"<<rmp[i]<<" ";
            for(auto i:result)cout<<" = "<<i<<" ";
            break;
        }
    }
	return 0;
}

File: Competitive Codes/ESKO_2020/esko.js
process.stdin.resume();
process.stdin.setEncoding('utf8');
var stdin_input = "";

process.stdin.on("data", function (input) {
    stdin_input += input;                               
    // Reading input from STDIN
});

process.stdin.on("end", function () {
   main(stdin_input);
});



class graph{
    constructor(){
        this.AdjList = new Map();
        this.mp  = {};
        this.rmp = {};
        this.mp_pro = [];
        this.result = [];
        this.flag = false;
    }
    calc_result(startingNode){
        this.result.sort((a,b)=>{
            return b[0]<a[0];
        });
        if(this.result[0][0]!==0){
            this.flag = true;
            var reso = "1";
            reso += startingNode;
            for(var i in this.result){
                reso += " = ";
                reso += this.result[i][0];
                reso += this.result[i][1];
            }
            console.log(reso);
        }
    }
    dochange(total_nodes){
        this.result=[];
        for(var i=0;i<total_nodes;i++){
            this.mp_pro[i] = 1;
        }
    }
    addVertex(vertex){
         this.AdjList.set(vertex, []);
    }
    addEdge(vertex, node, weight){
        this.AdjList.get(vertex).push({
            "dest":node,
            "weight":Number(weight),
            "reverse":false
        });
        this.AdjList.get(node).push({
            "dest":vertex,
            "weight":Number(weight),
            "reverse":true
        });
    }
    printGraph(){
        var get_keys = this.AdjList.keys();  
        for (var i of get_keys) { 
            var get_values = this.AdjList.get(i); 
            var conc = ""; 
            for (var j of get_values)
                console.log(j);
            console.log(i + " -> " + conc); 
        } 
    }
    mapper(point){
        for(var i=0;i<point.length;i++){
            this.mp[point[i]] = i;//["pallet"]=>0
            this.rmp[i] = point[i];//0=>"pallet"
        }
    }
    dfs(s,visited){
        visited[this.mp[s]] = true;
        var getNeighbors = this.AdjList.get(s);
        for(var i in getNeighbors){
            var get_dest_node = getNeighbors[i];
            if(!visited[this.mp[get_dest_node.dest]]){
                //calculation
                if(get_dest_node.reverse){
                    this.mp_pro[this.mp[get_dest_node.dest]] = this.mp_pro[this.mp[s]] /  get_dest_node.weight;
                    if(this.mp_pro[this.mp[get_dest_node.dest]]<1)
                        this.mp_pro[this.mp[get_dest_node.dest]] = 0;
                }else{
                    this.mp_pro[this.mp[get_dest_node.dest]] = this.mp_pro[this.mp[s]] *  get_dest_node.weight;
                }
                this.result.push([this.mp_pro[this.mp[get_dest_node.dest]],get_dest_node.dest]);
                this.dfs(get_dest_node.dest,visited);
            }
        }
    }
}


function main(input){
    var data = input.split('\n');
    const n = data.length;
    const point = data[0].split(';');
    node = new graph();
    for(var ii=0;ii< point.length;ii++){
        node.addVertex(point[ii]);
    }
    var decre = n-1;
    var k = 1;
    while(decre--){
        var ss = data[k].split(' ');k++;
        node.addEdge(ss[0],ss[3],ss[2]);
    }
    
    node.mapper(point);
    for(var ki=0;ki<n;ki++){
        if(node.flag) break;
        var visited = [];
        for(var ui=0;ui<n;ui++) visited[ui] = false;
        node.dochange(n);
        var startingNode = point[ki];
        node.dfs(startingNode,visited);
        node.calc_result(startingNode);
    }
}